text/0000-cfg-path-version.md

```diff
@@ -0,0 +1,956 @@
+while also being short enough to remain ergonomic to use.
+In particular, while `path_accessible` might be somewhat more unambiguous,
+we argue that from the context of seeing `accessible(::std::foo::bar)`
+it is clear that it is paths we are talking about because the argument
+is a path and not something else.
+
+#### `reachable`
+
+The word `reachable` is also a synonym of `accessible` and is one character 
+shorter. However, it tends to have a different meaning in code. Examples include:
+
++ `std::hint::unreachable_unchecked`
++ `std::unreachable`
+
+All in all, we have chosen to go with `accessible` instead as the
+more intuitive option.
+
+#### `usable`
+
+While `can_use` and `usable` are also strong contenders, we reject these options
+because they may imply to the user that only things that you may `use $path;` can
+go in there. Meanwhile, you may `#[cfg(accessible(::foo::MyTrait::my_method))`
+which is *not* possible as `use ::foo::MyTrait::my_method;`. This also applies
+to other associated items and inherent methods as well as `struct` fields.
+
+#### `has_path`
+
+Another strong contender is `has_path` or `have_path`.
+
+However, this variant is vague with respect to what "having" something means.
+In other words, it does not say whether it refers to being accessible and public,
+or whether it is usable, and so on.
+
+As we previously noted, having `path` in the
+name is also somewhat redundant because it is clear that `::std::bar` is a path.
+
+Another small wrinkle is that it is unclear whether it should be `have` or `has`.
+That choice depends on what one things the subject is. For example, if one 
+considers a module to be an "it", then it should probably be `has`.
+
+One upside to `has_path` is that it has precedent from the `clang` compiler.
+For example, a user may write: `#if __has_feature(cxx_rvalue_references)`
+or `__has_feature(c_generic_selections)`.
+
+Another benefit is that `has_` gives us the opportunity to introduce a family
+of `has_path`, `has_feature`, and `has_$thing` if we so wish.
+
+## `#[cfg(version(..))`
+
+When it comes to `version(..)`, it is needed to support conditional compilation
+of language features as opposed to library features as previously shown.
+Also, as we've seen, `version(..)` does not subsume `accessible(..)` but is
+rather a complementary mechanism.
+
+One problem specific to `version(..)` is that it might get too `rustc` specific.
+It might be difficult for other Rust implementations than `rustc` to work with
+this version numbering as libraries will compile against `rustc`s release
+numbering. However, it is possible for other implementations to follow
+`rustc` in the numbering and what features it provides. This is probably not
+too unreasonable as we can expect `rustc` to be the reference implementation
+and that other ones will probably lag behind. Indeed, this is the experience
+with `GHC` and alternative Haskell compilers.
```

Comment:
A drawback of this approach is that using `#[cfg(version(1.28))]` would make libsyntax fail to parse this syntax on all pre-existing toolchains to the stabilization of this feature. 

That is, if you have to support toolchains older than whatever version this attribute is this shipped with, you need to put the `#[cfg(version(...))]` behind a feature gated module like this, to avoid older toolchains from trying to parse it:

```
// The module where you want to use version
cfg_if! {
    if #[some_flag_detecting_compiler_version_in_build_rs] {
        mod internal;
        use internal::*;
    }
}

// and then in a separate internal.rs file
#[cfg(version(1.42))]
struct Foo;
```

Reply:
> and then have a build.rs that detects the rust toolchain version and sets --cfg toolchain_supports_cfg_version.

I mean, at that point, we can just implement this as a library that can be easily used from `build.rs` to define such config macros, which would allow the code above to just be written as: 

```rust
// In the module where you actually want to use `#[cfg(version)]`
cfg_if! {
    if #[cfg(rust_ge_1_42_0)] {
        struct Foo;
    }
}
```

Reply:
Note that the following syntax does not have this problem and is IIRC backward compatible down to Rust 1.0:

```rust
#[cfg(rust_version_ge = "1.42.0")]
struct Foo;
```

Reply:
@gnzlbg I think everyone considering using this feature understands that it will only work going *forward*, in versions that support `version` and `accessible`. That's still quite useful.

Reply:
@joshtriplett when it comes to older versions, there's a difference between "doesn't work" and "can't be parsed". I'd like to avoid the second situation if possible.

Reply:
I agree that it would still be useful. I was just pointing out a problem that I didn't see mentioned before. 

I'd like to use this feature in `libc`, and I won't be able to use it on new toolchains, because the old toolchains it must support cannot parse it. I sadly also know of many other crates that are still stuck supporting ~1.20.0 toolchains due to LTS linux distros. It will be a long time till these crates can start using this feature for detecting newer compiler toolchains due to the parsing incompatibilities.

I don't know if this problem is worth solving, particularly given that there is a workaround that works via the module file. Also even if it was worth solving, I don't know if it would be worth it to solve this here, since the feature is already useful as is. 

If we are going to make the effort of implementing this, one possibility would be to introduce a backward compatible syntax, e.g., in Rust2015 and deprecated in Rust2018, that we can remove in the next edition or something. Crates stuck supporting older toolchains are stuck on older editions anyways. Don't know if this would be worth it.

Reply:
We initially started with `version = ".."` for this reason and switched cause @eddyb thought it looked too much like a static "lookup" flag rather than a "dynamic" flag. I'm inclined to agree with that reasoning but it's a good point re. older versions. This is discussed a bit in the RFC text wrt. the rationale for the current syntax.

We haven't dug that much into the syntax, but it's a good discussion to have (which can be done mostly on the tracking issue as I consider these to be primarily details).

Reply:
This does bring though the idea that whichever release `version` becomes stable will probably become the new minimum stable rust version that a lot of libraries will bump to. It might be good to pair it's release with a major feature(s) so that the most amount of people can be benefit from them.

Reply:
It's probably not that difficult to support two versions; it's mostly just a bit of parsing inside or outside the string and they'll be uniform otherwise.

Reply:
@XAMPPRocky that's pretty clever =P will just have to find some major feature to couple with heh.

Reply:
> This does bring though the idea that whichever release version becomes stable will probably become the new minimum stable rust version that a lot of libraries will bump to.

At best, depending on how `cfg(version)` is shipped, projects might be able to just use it for detecting newer features, but it will live alongside the feature-detection code that these projects are already using. The "old" feature detection code is simple and low maintenance, so bumping the MSRV to be able to exclusively use `cfg(version)` is something that doesn't add much value, while compromising something that's important for these project: the only reason a project would do feature detection is to be able to support older toolchains, so the project must see value in that.

I think that maybe eventually all projects might be able to just use `cfg(version)`, but that will be a consequence of them bumping the MSRV due to other issues, and not because there is a lot of value in bumping the MSRV to a toolchain that supports `cfg(version)`.

Reply:
> At best, depending on how cfg(version) is shipped, projects might be able to just use it for detecting newer features

I don't think `cfg(version)` has to be immediately valuable. Eventually more features will be covered under `cfg(version)` than those features that aren't, and then it will be more compelling for more libraries.

> so bumping the MSRV to be able to exclusively use cfg(version) is something that doesn't add much value

Unless your MSRV is one version behind `cfg(version)` your value won't be exclusively `cfg(version)`. If your MSRV is 1.15 for example and `cfg(version)` was released in 1.40.0 you'd be getting everything from 1.16.0–1.40.0, and you'd be able to use `cfg(version)` to ease future feature development.